.github/workflows/alarm.yml

```diff
@@ -18,9 +21,364 @@ jobs:
+                  if k in drop or v is None: continue
+                  k = str(k).strip(); v = str(v).strip()
+                  if not k or not v: continue
+                  pairs.append((k, v))
+              pairs.sort(key=lambda x: x[0])
+              return "-".join(f"{k}={v}" for k, v in pairs)
+
+          def wandb_normalize_benchmark_id(bid: str) -> str:
+              kv = {}
+              for token in (p.strip() for p in bid.split("-") if p.strip()):
+                  if "=" not in token: continue
+                  k, v = token.split("=", 1)
+                  kv[k.strip()] = v.strip()
+              return _normalize_kv_id(kv)
+
+          def artifacts_parse_speed_txt_lines(lines):
+              METRIC_KEYS = {"compile_time", "runtime_fps", "realtime_factor"}
+              out = {}
+              for line in lines:
+                  if not line.strip(): continue
+                  parts = [p.strip() for p in line.strip().split("|") if "=" in p]
+                  kv = {}
+                  for p in parts:
+                      k, v = p.split("=", 1)
+                      kv[k.strip()] = v.strip()
+                  tid = _normalize_kv_id(kv, drop_keys=METRIC_KEYS)
+                  rt = kv.get("runtime_fps"); ct = kv.get("compile_time")
+                  try: rt = float(rt) if rt is not None else None
+                  except: rt = None
+                  try: ct = float(ct) if ct is not None else None
+                  except: ct = None
+                  out[tid] = {"runtime_fps": rt, "compile_time": ct}
+              return out
+
+          def fmt_num(v):
+              if v is None or (isinstance(v,float) and math.isnan(v)): return "—"
+              try:
+                  if isinstance(v, float) and not v.is_integer():
+                      return f"{v:.2f}"
+                  return f"{int(v):,}"
+              except Exception:
+                  return str(v)
+
+          def fmt_pct(v, highlight=False):
+              if v is None or (isinstance(v,float) and math.isnan(v)): return "—"
+              s = f"{v:+.2f}%"
+              return f"**{s}**" if highlight else s
+
+          # ----- thresholds -----
+          tol_rt = float(os.environ.get("RUNTIME_REGRESSION_TOLERANCE_PCT","10"))
+          tol_ct = float(os.environ.get("COMPILE_REGRESSION_TOLERANCE_PCT","10"))
+          MAX_REVISIONS = int(os.environ.get("MAX_REVISIONS","5"))
+          NO_CHANGE_PATIENCE = int(os.environ.get("NO_CHANGE_PATIENCE","100"))
+
+          # ----- load artifact (current results) -----
+          artifacts_path = os.path.abspath(os.environ.get("ARTIFACTS_DIR", "./artifacts"))
+          if not os.path.exists(artifacts_path):
+              open(os.environ.get("PR_COMMENT_PATH","pr_comment.md"),"w").close()
+              open(os.environ.get("CHECK_BODY_PATH","check_output.md"),"w").close()
+              sys.exit(0)
+
+          current_txt_path = None
+          for root, _, files in os.walk(artifacts_path):
```

Comment:
Use `pathlib` instead of `os`.

You should load all these path arguments from env variables as pathlib.Path objects directly.